# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaCommon.h
# Commit: 2fb182664097
# Full Hash: 2fb182664097f1b22021eec844d0d1adbffc40dc
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-12-05 09:38:58
# Regressor Bug: 1678030
# File Overlap Count: 1
# Description:
#   Bug 1678030 - Support multiple variants of hasResult expectations in CreateAndExecuteSingleStepStatement. r=dom-workers-and-storage-reviewers,ttung
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D97461
# ==============================================================================

diff -r 2ed5e139d784 -r 2fb182664097 dom/quota/QuotaCommon.h
--- a/dom/quota/QuotaCommon.h	Fri Dec 04 14:56:18 2020 +0000
+++ b/dom/quota/QuotaCommon.h	Fri Dec 04 15:10:07 2020 +0000
@@ -979,10 +979,26 @@
 Result<nsCOMPtr<nsIFile>, nsresult> CloneFileAndAppend(
     nsIFile& aDirectory, const nsAString& aPathElement);
 
+enum class SingleStepResult { AssertHasResult, ReturnNullIfNoResult };
+
+template <SingleStepResult ResultHandling>
+using SingleStepSuccessType =
+    std::conditional_t<ResultHandling == SingleStepResult::AssertHasResult,
+                       NotNull<nsCOMPtr<mozIStorageStatement>>,
+                       nsCOMPtr<mozIStorageStatement>>;
+
 // Creates a statement with the specified aStatementString, executes a single
-// step, and asserts that it has a result, and returns the statement. Any other
-// errors are propagated.
-Result<nsCOMPtr<mozIStorageStatement>, nsresult>
+// step, and returns the statement.
+// Depending on the value ResultHandling,
+// - it is asserted that there is a result (default resp.
+//   SingleStepResult::AssertHasResult), and the success type is
+//   MovingNotNull<nsCOMPtr<mozIStorageStatement>>
+// - it is asserted that there is no result, and the success type is Ok
+// - in case there is no result, nullptr is returned, and the success type is
+//   nsCOMPtr<mozIStorageStatement>
+// Any other errors are always propagated.
+template <SingleStepResult ResultHandling = SingleStepResult::AssertHasResult>
+Result<SingleStepSuccessType<ResultHandling>, nsresult>
 CreateAndExecuteSingleStepStatement(mozIStorageConnection& aConnection,
                                     const nsACString& aStatementString);
 
